### PR TITLE
topic reflection

### PR DESCRIPTION
**Pull request by RanMaoyi**
After https://github.com/nanomq/NanoNNG/pull/692 and sync nng

### CODE REVIEW

**Review by lee-emqx — APPROVED** [review]

File: docs/en_US/http-api/v4.md [hunk 1]
```diff
@@ -1049,11 +1078,13 @@ $ curl --location 'http://127.0.0.1:8081/api/v4/bridges/sub/emqx' \
...
             {
-                "topic": "cmd/topic4",
+                "remote_topic": "cmd/topic4",
+                "local_topic": "local/topic4",
                 "qos": 1
             },
             {
-                "topic": "cmd/topic5",
+                "remote_topic": "cmd/topic5",
+                "local_topic": "cmd/topic5",
```
Comment (wanghaEMQ) [inline docs/en_US/http-api/v4.md:1087]: local/topic5

File: nanomq/apps/broker.c [hunk 2]
```diff
@@ -168,31 +167,41 @@ bridge_handler(nano_work *work)
 		    &props, work->pub_packet->var_header.publish.properties);
 	}
```
Comment (wanghaEMQ) [inline nanomq/apps/broker.c:168]: Might memory leakage in props?

File: nanomq/bridge.c [hunk 3]
```diff
@@ -127,6 +127,43 @@ create_disconnect_msg()
...
+						work->pub_packet->var_header.publish.topic_name.len = sub_topic->local_topic_len;
+
+						return 0;
+					}
+				}
+			}
+		}
+	}
+	return -1;
+}
```
Comment (wanghaEMQ) [inline nanomq/bridge.c:165]: Too much folding...

File: nanomq/bridge.c [hunk 4]
```diff
@@ -127,6 +127,43 @@ create_disconnect_msg()
...
+				if (topic != NULL && node->sub_list[i]->remote_topic != NULL) {
+					if (strncmp(topic, node->sub_list[i]->remote_topic, len) == 0) {
+						topics *sub_topic = node->sub_list[i];
+
+						char *local_topic = nng_alloc(sub_topic->local_topic_len + 1);
+						if (local_topic == NULL) {
+							log_error("bridge: alloc local_topic failed");
+							return -1;
+						}
+						memset(local_topic, 0, sub_topic->local_topic_len + 1);
```
Comment (wanghaEMQ) [inline nanomq/bridge.c:148]: Why not try nng_strdup or strdup?

File: nanomq/bridge.c [hunk 5]
```diff
@@ -127,6 +127,43 @@ create_disconnect_msg()
...
+bridge_handle_sub_reflection(nano_work *work, conf_bridge *bridge)
+{
+	for (size_t t = 0; t < bridge->count; t++) {
+		conf_bridge_node *node = bridge->nodes[t];
+		if (node->enable) {
+			for (size_t i = 0; i < node->sub_count; i++) {
+				char *topic = work->pub_packet->var_header.publish.topic_name.body;
+				int len = work->pub_packet->var_header.publish.topic_name.len;
+				if (topic != NULL && node->sub_list[i]->remote_topic != NULL) {
+					if (strncmp(topic, node->sub_list[i]->remote_topic, len) == 0) {
```
Comment (wanghaEMQ) [inline nanomq/bridge.c:140]: What if the topic is the prefix of the remote_topic.
My suggestion. Use strcmp rather than strncmp.

**Review by wanghaEMQ — COMMENTED** [review]
In addition to the comments above.

Need more tests.
Did you benchmark the functionalities that would be affected?

And for the name of local/fwd/cmd. Maybe we need more talk with @JaylinYu.
They are a little bit hard to tell.

**Comment by JaylinYu** [comment]
Better use msg dup

**Comment by RanMaoyi** [comment]
> Better use msg dup.
> Will have a deep look tmw

I took a look at the implementation of msg_dup and found that msg_dup would be alloc multiple times, causing a noticeable performance degradation

The current approach is to modify the packet topic information, which is simple and efficient

So both schemes have their advantages and disadvantages

If there is any benefit that I don't notice that only using msg_dup yields, wish to point it out

**Comment by RanMaoyi** [comment]
@wanghaEMQ Thanks for your suggestions, which are helpful.

I fix them one by one, please check again :P

**Pull request merged**